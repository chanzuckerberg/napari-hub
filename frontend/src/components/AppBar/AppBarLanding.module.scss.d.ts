export type Styles = {
  heading: string;
  list: string;
  logo: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
