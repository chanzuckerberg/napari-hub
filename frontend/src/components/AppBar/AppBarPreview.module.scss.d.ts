export type Styles = {
  fieldInfo: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
