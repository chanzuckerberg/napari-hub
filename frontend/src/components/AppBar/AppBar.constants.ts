import { LINKS } from '@/constants';
import { LinkInfo } from '@/types';

export const APP_LINKS: LinkInfo[] = [LINKS.ABOUT, LINKS.FAQ];
