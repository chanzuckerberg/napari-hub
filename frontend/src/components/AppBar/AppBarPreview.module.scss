@import '@/utils/styles';

.fieldInfo {
  padding: calculate-rem(6px);
}
