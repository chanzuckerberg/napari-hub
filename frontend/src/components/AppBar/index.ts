export * from './AppBar';
export * from './AppBarLanding';
