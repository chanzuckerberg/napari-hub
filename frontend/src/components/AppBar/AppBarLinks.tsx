import clsx from 'clsx';

import { I18n } from '@/components/I18n';
import { Link } from '@/components/Link';
import { resetLoadingState } from '@/store/loading';
import { LinkInfo } from '@/types';

interface CommonProps {
  vertical?: boolean;
}

interface LinkListProps extends CommonProps {
  items: LinkInfo[];
}

/**
 * Link bar for rendering menu links. This only shows up on lg+ screens.
 */
function LinkList({ items, vertical }: LinkListProps) {
  return (
    <div className={clsx('flex', vertical && 'flex-col')}>
      {items.map((item) => (
        <Link
          className={clsx(vertical ? 'mt-sds-l' : 'ml-sds-xl')}
          key={item.link}
          href={item.link}
          onClick={resetLoadingState}
        >
          {item.title}
        </Link>
      ))}
    </div>
  );
}

interface Props extends CommonProps {
  className?: string;
  items?: LinkInfo[];
}

/**
 * Component that includes site-wide navigational links.
 */
export function AppBarLinks({ className, items, vertical }: Props) {
  return (
    <nav className={clsx(className, 'flex', vertical && 'flex-col')}>
      <Link
        data-testid="appBarHome"
        className="whitespace-nowrap"
        // Redirect to home page
        href="/"
        onClick={resetLoadingState}
      >
        <I18n i18nKey="common:napariHub.strong" />
      </Link>

      {items && <LinkList items={items} vertical={vertical} />}
    </nav>
  );
}
