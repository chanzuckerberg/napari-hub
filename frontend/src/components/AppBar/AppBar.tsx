import IconButton from '@material-ui/core/IconButton';
import clsx from 'clsx';
import { useTranslation } from 'next-i18next';
import { useRef, useState } from 'react';

import { ColumnLayout } from '@/components/ColumnLayout';
import { Menu } from '@/components/icons';
import { Media } from '@/components/media';
import { MenuPopover } from '@/components/MenuPopover';
import { SearchBar } from '@/components/SearchBar';

import { AppBarLinks } from './AppBarLinks';
import { useAppBarLinks } from './useAppBarLinks';

/**
 * App bar component that renders the home link, search bar, and menu.
 */
export function AppBar() {
  const anchorElRef = useRef<HTMLButtonElement | null>(null);
  const [visible, setVisible] = useState(false);
  const links = useAppBarLinks();
  const [t] = useTranslation(['common']);

  return (
    <>
      <Media lessThan="screen-600">
        <MenuPopover
          anchorEl={anchorElRef.current}
          items={links}
          onClose={() => setVisible(false)}
          visible={visible}
        />
      </Media>

      <ColumnLayout
        className={clsx(
          // Color and height
          'bg-napari-primary h-napari-app-bar',

          // Centering
          'justify-center items-center',

          // Padding
          'px-6 screen-495:px-12 screen-1150:p-0',

          // Grid layout for smaller screens. This allows the search bar to
          // extend to its max width to the left. The `zero:` modifier is used
          // to increase specificity over ColumnLayout.
          'zero:grid-cols-[min-content,1fr]',
        )}
        component="header"
      >
        <Media greaterThanOrEqual="screen-600">
          <AppBarLinks items={links} />
        </Media>

        <Media lessThan="screen-600">
          <AppBarLinks />
        </Media>

        <div
          className={clsx(
            // Flex layout
            'flex items-center',

            // Take 100% of width.
            'w-full',

            // Align container to the right of the grid cell
            'justify-self-end',

            // Use more columns on larger screens
            'screen-875:col-span-2 screen-1150:col-span-3',
          )}
        >
          <SearchBar />

          {/* Menu button */}
          <Media className="ml-6 flex" lessThan="screen-600">
            <IconButton onClick={() => setVisible(true)} ref={anchorElRef}>
              <Menu alt={t('common:alt.sideMenu')} />
            </IconButton>
          </Media>
        </div>
      </ColumnLayout>
    </>
  );
}
