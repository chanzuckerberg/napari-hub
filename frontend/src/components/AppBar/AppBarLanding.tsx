import clsx from 'clsx';
import { useTranslation } from 'next-i18next';
import { ReactNode } from 'react';

import { ColumnLayout } from '@/components/ColumnLayout';
import { Hub } from '@/components/icons';
import { Media, MediaFragment } from '@/components/media';

import styles from './AppBarLanding.module.scss';
import { AppBarLinks } from './AppBarLinks';
import { useAppBarLinks } from './useAppBarLinks';

interface ListProps {
  children: ReactNode;
  className?: string;
}

function List({ children, className }: ListProps) {
  return (
    <ul
      className={clsx(
        'list-disc font-semibold space-y-4',
        styles.list,
        className,
      )}
    >
      {children}
    </ul>
  );
}

/**
 * Component that renders the landing page variant of the AppBar.
 */
export function AppBarLanding() {
  const links = useAppBarLinks();
  const [t] = useTranslation(['homePage']);

  // On smaller layouts, the last list item is rendered below the icon and 2
  // preceding list items. To achieve this affect, the last list item is
  // rendered in a separate list so that it can be rendered below everything.
  const lastListNode = <li>{t('homePage:appBar.share')}</li>;

  return (
    <ColumnLayout
      className="bg-napari-primary p-6 screen-495:p-12"
      classes={{
        // Use 3-column layout instead of 4-column.
        fourColumn: 'screen-1150:grid-cols-napari-3',
        gap: 'gap-x-6 md:gap-x-12',
      }}
      component="header"
    >
      <Media greaterThanOrEqual="screen-1425">
        <AppBarLinks items={links} vertical />
      </Media>

      <Media lessThan="screen-1425">
        <AppBarLinks className="mb-6" items={links} />
      </Media>

      <h1
        className={clsx(
          'font-bold col-span-full',
          'mb-6 screen-495:mb-12',
          'screen-1425:col-start-2 screen-1425:col-span-3',
          styles.heading,
        )}
      >
        {t('homePage:appBar.title')}
      </h1>

      <div
        className={clsx(
          'grid grid-cols-[min-content,1fr]',
          'gap-6 screen-655:gap-12',
          'col-span-2 screen-875:col-span-3',
          'screen-1425:col-start-2',
          'items-center',
        )}
      >
        <Hub className={styles.logo} />

        <List className="ml-7">
          <li>{t('homePage:appBar.discover')}</li>
          <li>{t('homePage:appBar.learnHow')}</li>

          {/* Render in list if screen is wide enough. */}
          <MediaFragment greaterThanOrEqual="screen-725">
            {lastListNode}
          </MediaFragment>
        </List>
      </div>

      {/* Render as separate list so that it renders below everything on smaller screens. */}
      <MediaFragment lessThan="screen-725">
        <List className="col-span-2 px-6 mt-4">{lastListNode}</List>
      </MediaFragment>
    </ColumnLayout>
  );
}
