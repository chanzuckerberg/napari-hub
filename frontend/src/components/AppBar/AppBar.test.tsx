import { render } from '@testing-library/react';

import { SearchStoreProvider } from '@/store/search/context';

import { AppBar } from './AppBar';

describe('<AppBar />', () => {
  it('should match snapshot', () => {
    const component = render(
      <SearchStoreProvider>
        <AppBar />
      </SearchStoreProvider>,
    );
    expect(component.asFragment()).toMatchSnapshot();
  });
});
