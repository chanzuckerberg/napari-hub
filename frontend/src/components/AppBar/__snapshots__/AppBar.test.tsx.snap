// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppBar /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="screen-600:hidden"
  />
  <header
    class="bg-napari-primary h-napari-app-bar justify-center items-center px-6 screen-495:px-12 screen-1150:p-0 zero:grid-cols-[min-content,1fr] grid justify-center gap-6 screen-495:gap-12 grid-cols-2 screen-875:grid-cols-napari-3 screen-1150:grid-cols-napari-4 screen-1425:grid-cols-napari-5"
  >
    <div
      class="hidden screen-600:block"
    >
      <nav
        class="flex"
      >
        <a
          class="whitespace-nowrap"
          data-testid="appBarHome"
          href="/"
        >
          napari &lt;strong&gt;hub&lt;/strong&gt;
        </a>
        <div
          class="flex"
        >
          <a
            class="ml-6"
            href="/about"
          >
            About
          </a>
          <a
            class="ml-6"
            href="/faq"
          >
            FAQ
          </a>
        </div>
      </nav>
    </div>
    <div
      class="screen-600:hidden"
    >
      <nav
        class="flex"
      >
        <a
          class="whitespace-nowrap"
          data-testid="appBarHome"
          href="/"
        >
          napari &lt;strong&gt;hub&lt;/strong&gt;
        </a>
      </nav>
    </div>
    <div
      class="flex items-center w-full justify-self-end screen-875:col-span-2 screen-1150:col-span-3"
    >
      <form
        class="flex flex-auto items-center border-b-2 border-black"
        data-testid="searchBarForm"
      >
        <input
          aria-label="Search bar for searching for napari plugins."
          class="flex flex-auto border-none outline-none bg-transparent w-0"
          data-testid="searchBarInput"
          value=""
        />
        <button
          aria-label="Submit search query"
          class="MuiButtonBase-root MuiIconButton-root"
          data-testid="submitQueryButton"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              class="h-5 w-5"
              fill="none"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.32004 7.68001L0.746704 15.2533M10.6667 9.60001C13.0134 9.60001 14.9334 7.68001 14.9334 5.33335C14.9334 2.98668 13.0134 1.06668 10.6667 1.06668C8.32004 1.06668 6.40004 2.98668 6.40004 5.33335C6.40004 7.68001 8.32004 9.60001 10.6667 9.60001Z"
                stroke="black"
                stroke-width="2"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </form>
      <div
        class="ml-6 flex screen-600:hidden"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              fill="none"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                Icon for opening side menu.
              </title>
              <path
                d="M16 1H0"
                stroke="black"
                stroke-width="1.98857"
              />
              <path
                d="M16 7.99924H0"
                stroke="black"
                stroke-width="1.98857"
              />
              <path
                d="M16 15H0"
                stroke="black"
                stroke-width="1.98857"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
